include/seqan/blast/blast_io_context.h

```diff
@@ -0,0 +1,383 @@
+
+#ifndef SEQAN_BLAST_BLAST_IO_CONTEXT_H__
+#define SEQAN_BLAST_BLAST_IO_CONTEXT_H__
+
+namespace seqan
+{
+
+//forward declare
+template <typename TScore>
+struct BlastScoringScheme;
+
+/*!
+ * @class BlastIOContext
+ * @headerfile <seqan/blast.h>
+ * @signature template <typename TScore_ = Blosum62, typename TString = std::string,
+ *           BlastProgram p = BlastProgram::UNKNOWN, BlastTabularSpec h = BlastTabularSpec::UNKNOWN>
+ * struct BlastIOContext { ... };
+ *
+ * @brief An object that holds file global information and buffers for BlastIO
+ *
+ * @tparam TScore   Type of the @link Score @endlink object used.
+ * @tparam TString  Type of the @link StringConcept @endlink used for members and buffers.
+ * @tparam p        @link BlastProgram @endlink as compile-time parameter.
+ * @tparam h        @link BlastTabularSpec @endlink as compile-time parameter.
+ *
+ * This needs to be passed to most read*(), skip*() and write*() functions as
+ * a parameter. Before writing, some of the context's members should be set; after reading it will contain
+ * all information from the file that did not belong to a @link BlastRecord @endlink, e.g. the name of the database.
+ * It also contains buffers for internal use.
+ *
+ * You should re-use this object (i.e. only create it once for
+ * every file that you read/write). And you don't need to and should not clear()
+ * this, except when restarting IO on a different file.
+ *
+ * To speed-up file writing slightly you can set the value template parameters <tt>p</tt> and/or <tt>h</tt> to something
+ * other than ::UNKNOWN at compile-time (e.g. if you know that you will be printing only BLASTX), but then you won't
+ * be able to modify these values with @link BlastIOContext#setBlastProgram @endlink and
+ * @link BlastIOContext#setBlastTabularSpec @endlink at run-time. For file reading this also possible, but usually the
+ * added flexibility of automatically detecting these values is prefferable.
+ *
+ * @section Example
+ *
+ * Here as an example of which members to set on a context, before using it for Output:
+ * @code{.cpp}
+ * typedef BlastIOContext<Blosum62> TContext;
+ * BlastReportOut<TContext> outfile("/tmp/output.blast");
+ *
+ * // set gap parameters in blast notation
+ * setScoreGapOpenBlast(context(outfile).scoringScheme, -11);
+ * setScoreGapExtend(context(outfile).scoringScheme, -1);
+ *
+ * // protein vs protein search is BLASTP
+ * setBlastProgram(context(outfile), BlastProgram::BLASTP);
+ *
+ * // set the database properties in the context
+ * context(outfile).dbName = "The Foo Database";
+ * context(outfile).dbTotalLength = length(concat(subjects));
+ * context(outfile).dbNumberOfSeqs = length(subjects);
+ * @endcode
+ *
+ * See @link BlastTabularOut @endlink and @link BlastReportOut @endlink for more complete examples of usage.
+ */
+
+template <typename TScore_ = Blosum62,
+          typename TString_ = std::string,
```

Comment:
Not important. Using std::string instead of CharString is a little more efficient in some places, but I though maybe someone would want to use CharString for some reasons.